src/shogun/clustering/KMeansLloydImpl.cpp

```diff
@@ -17,19 +18,30 @@ using namespace shogun;
 
 namespace shogun
 {
-void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_iter, SGMatrix<float64_t> mus,
-		SGVector<int32_t> ClList, SGVector<float64_t> weights_set, bool fixed_centers)
+void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_iter, SGMatrix<float64_t> mus, 
+		bool fixed_centers)
 {
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
-	int32_t XSize=lhs->get_num_vectors();
+	
+	int32_t lhs_size=lhs->get_num_vectors();
 	int32_t dimensions=lhs->get_num_features();
 
 	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
 	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
 
-	SGVector<float64_t> dists=SGVector<float64_t>(k*XSize);
-	dists.zero();
+	/* Cluster assignments for each point */
+	SGVector<int32_t> cl_list=SGVector<int32_t>(lhs_size);
+	cl_list.zero();
+
+	/* Weights : Number of points in each cluster */
+	SGVector<float64_t> weights_set=SGVector<float64_t>(k);
+	weights_set.zero();
+	/* Initially set all weights for zeroth cluster, Changes in assignement step */
+	weights_set[0]=lhs_size;
+
+	if (distance->get_distance_type() == D_EUCLIDEAN)
```

Comment:
this is not a good idea.
Sorry I know we had a lengthy discussion, but I did not realise this.
The KMeans class should not use any specific features of the distance class specialization. Rather but a virtual method "precompute" in the base class